janusgraph-core/src/main/java/org/janusgraph/graphdb/tinkerpop/optimize/JanusGraphMixedIndexCountStep.java

```diff
@@ -0,0 +1,119 @@
+import org.apache.tinkerpop.gremlin.structure.util.StringFactory;
+import org.janusgraph.core.JanusGraphElement;
+import org.janusgraph.core.JanusGraphIndexQuery;
+import org.janusgraph.core.JanusGraphTransaction;
+import org.janusgraph.core.MixedIndexCountQuery;
+import org.janusgraph.graphdb.internal.ElementCategory;
+import org.janusgraph.graphdb.internal.OrderList;
+import org.janusgraph.graphdb.query.JanusGraphPredicateUtils;
+import org.janusgraph.graphdb.query.QueryUtil;
+import org.janusgraph.graphdb.query.condition.Condition;
+import org.janusgraph.graphdb.query.condition.MultiCondition;
+import org.janusgraph.graphdb.query.condition.PredicateCondition;
+import org.janusgraph.graphdb.query.graph.IndexQueryBuilder;
+import org.janusgraph.graphdb.query.graph.JointIndexQuery;
+import org.janusgraph.graphdb.query.graph.MixedIndexCountQueryBuilder;
+import org.janusgraph.graphdb.query.index.IndexSelectionStrategy;
+import org.janusgraph.graphdb.query.index.SingleMixedIndexSelectionStrategy;
+import org.janusgraph.graphdb.query.profile.QueryProfiler;
+import org.janusgraph.graphdb.tinkerpop.profile.TP3ProfileWrapper;
+import org.janusgraph.graphdb.transaction.StandardJanusGraphTx;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Objects;
+import java.util.Set;
+
+/**
+ * @author Boxuan Li (liboxuan@connect.hku.hk)
+ */
+public class JanusGraphMixedIndexCountStep<S> extends ReducingBarrierStep<S, Long> implements Profiling {
+
+    private final List<HasContainer> hasContainers = new ArrayList<>();
+    private QueryProfiler queryProfiler = QueryProfiler.NO_OP;
+    private QueryProfiler optProfiler = QueryProfiler.NO_OP;
+    private MixedIndexCountQuery mixedIndexCountQuery;
+    private boolean done = false;
+    private IndexSelectionStrategy indexSelector = new SingleMixedIndexSelectionStrategy();
+
+    public JanusGraphMixedIndexCountStep(final Traversal.Admin traversal) {
+        super(traversal);
+    }
+
+    public void setMixedIndexCountQuery(MixedIndexCountQuery mixedIndexCountQuery) {
+        Objects.requireNonNull(mixedIndexCountQuery);
+        this.mixedIndexCountQuery = mixedIndexCountQuery;
+    }
+
+    public MixedIndexCountQuery buildGlobalGraphIndexQuery(final GraphStep originalGraphStep, final JanusGraphTransaction tx) {
+        optProfiler = new TP3ProfileWrapper(new MutableMetrics(this.getId() + "optimization", "optimization"));
+        MixedIndexCountQueryBuilder countQueryBuilder = (MixedIndexCountQueryBuilder) tx.mixedIndexCountQuery();
+        List<PredicateCondition<String, JanusGraphElement>> constraints = new ArrayList<>();
+        for (final HasContainer condition : hasContainers) {
+            constraints.add(new PredicateCondition<>(condition.getKey(), JanusGraphPredicateUtils.convert(condition.getBiPredicate()), condition.getValue()));
+        }
+        return countQueryBuilder.constructIndex(optProfiler, constraints,
```

Comment:
I like the Idea that the step itself performs the time consuming task of index selection, which makes it possible to track the work done in the optimization phase. Great idea!